openedx/core/djangoapps/content/course_overviews/models.py

```diff
@@ -0,0 +1,240 @@
+            has_active_web_certificates=(get_active_web_certificate(course) is not None),
+            cert_name_short=course.cert_name_short,
+            cert_name_long=course.cert_name_long,
+            lowest_passing_grade=course.lowest_passing_grade,
+            end_of_course_survey_url=course.end_of_course_survey_url,
+
+            mobile_available=course.mobile_available,
+            visible_to_staff_only=course.visible_to_staff_only,
+            _pre_requisite_courses_json=json.dumps(course.pre_requisite_courses)
+        )
+
+    @staticmethod
+    def get_from_id(course_id):
+        """Load a CourseOverview object for a given course ID.
+
+        First, we try to load the CourseOverview from the database. If it
+        doesn't exist, we load the entire course from modulestore, create a
+        CourseOverview object from it, and then cache it in the database for
+        future use.
+
+        Args:
+            course_id (CourseKey): the ID of the course overview to be loaded
+
+        Returns:
+            CourseOverview: overview of the requested course
+        """
+        course_overview = None
+        try:
+            course_overview = CourseOverview.objects.get(id=course_id)
+            # Cache hit! Just return the overview
+        except CourseOverview.DoesNotExist:
+            # Cache miss. Load entire course and create a CourseOverview from it
+            course = modulestore().get_course(course_id)
+            if course:
+                course_overview = CourseOverview._create_from_course(course)
+                course_overview.save()  # Save new overview to the cache
+        return course_overview
+
+    def __repr__(self):
+        """
+        Returns a simple string representation of this object for debugging.
+
+        Example return value: CourseOverview(location=course-v1://edX+DemoX.1+2014)
+        """
+        return "{}(location={})".format(self.__class__.__name__, self.location)
+
+    def __str__(self):
+        """
+        Returns a string representation of this object suitable for a user to see.
+        """
+        return unicode(self.id)
+
+    def clean_id(self, padding_char='='):
+        """
+        Returns a unique deterministic base32-encoded ID for the course.
+
+        Args
+            padding_char (str): Character used for padding at end of base-32
+                                -encoded string, defaulting to '='
+        """
+        return course_metadata_utils.clean_id(self.location, padding_char)
+
+    @property
+    def location(self):
+        """Returns the UsageKey of this course.
+
+        UsageKeyField has a strange behavior where it fails to parse the "run"
+        of a course out of the serialized form of a Mongo Draft UsageKey. This
+        method is a wrapper around _location attribute that fixes the problem
+        by calling map_into_course, which restores the run attribute.
+        """
+        return self._location.map_into_course(self.id)
```

Comment:
For performance reasons, move the call to `map_into_course` to the `_create_from_course` method.
`map_into_course` needs to lookup the course in the modulestore's mapping table, so it can be a costly overhead since callers of this property expect it to be instantaneous.